### PR TITLE
Preliminary allocation domain support

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stacked on https://github.com/NVIDIA/Fuser/pull/299 https://github.com/NVIDIA/Fuser/pull/298 https://github.com/NVIDIA/Fuser/pull/296, Split out from https://github.com/NVIDIA/Fuser/pull/276. ~Please wait until https://github.com/NVIDIA/Fuser/pull/299 https://github.com/NVIDIA/Fuser/pull/298 https://github.com/NVIDIA/Fuser/pull/296 merged before review.~ (Edit: rebased)

This PR modifies our existing indexing to bring preliminary allocation domain support. This allocation domain support is by no means complete, and my bar for this work is:
1. Newly added cases must be working
2. No existing cases fail

And I am using a minimum effort to make the above bar satisfied. In future PRs, I will do more iteration to make more things supported, until at some point I feel that it takes too much effort to modify existing indexing approach, at that time, I will stop working on this.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/index_compute.cpp [hunk 1]
```diff
@@ -1598,90 +1598,90 @@ std::vector<Val*> Index::getNonGlobalProducerStridedIndices(
   const auto& zero_domain_map = producer_indexing.zeroDomains();
   // Indices should now be mapped onto IterDomains in producer, so just grab
   // and use them.
-  auto root_dom = producer_tv->getMaybeRFactorDomain();
+  auto alloc_dom = producer_tv->getMaybeAllocationDomain();
```
Comment (naoyam) [inline csrc/index_compute.cpp:1601]: nit: const auto&

File: csrc/index_compute.cpp [hunk 2]
```diff
@@ -1733,22 +1733,22 @@ std::vector<Val*> Index::getProducerPerDimLogicalIndex(
...
+  auto guard = ir_utils::allocateToRFactorDomainGuard(producer_tv, false);
+  return getProducerAllocationIndices(
       producer_tv, consumer_tv, loops, rotated_loops, override_index);
 }
 
 std::vector<Val*> Index::getStrides(const TensorView* tv) {
   // Indices should now be mapped onto IterDomains in consumer, so just grab
   // and use them.
-  auto root_dom = tv->getMaybeRFactorDomain();
+  auto alloc_dom = tv->getMaybeAllocationDomain();
```
Comment (naoyam) [inline csrc/index_compute.cpp:1744]: const auto&

File: csrc/index_compute.cpp [hunk 3]
```diff
@@ -1779,60 +1779,61 @@ std::vector<Val*> Index::getStrides(const TensorView* tv) {
...
   return strides;
 }
 
-std::vector<Val*> Index::getConsumerRootIndices(
+std::vector<Val*> Index::getConsumerAllocationIndices(
     const TensorView* tv,
     const std::vector<kir::ForLoop*>& loops,
     const IndexFromIdGraph& index_from_id_graph) {
-  auto root_dom = tv->getMaybeRFactorDomain();
+  auto alloc_dom = tv->getMaybeAllocationDomain();
```
Comment (naoyam) [inline csrc/index_compute.cpp:1799]: const auto&

File: csrc/index_compute.cpp [hunk 4]
```diff
@@ -1900,53 +1901,53 @@ std::vector<Val*> Index::getProducerRootIndices(
 
   // Indices should now be mapped onto IterDomains in producer, so just grab
   // and use them.
-  auto root_dom = producer_tv->getMaybeRFactorDomain();
+  auto alloc_dom = producer_tv->getMaybeAllocationDomain();
```
Comment (naoyam) [inline csrc/index_compute.cpp:1904]: const auto&

File: csrc/index_compute.cpp [hunk 5]
```diff
@@ -2031,74 +2032,74 @@ std::vector<Val*> Index::getNonGlobalConsumerStridedIndices(
 
   // Indices should now be mapped onto IterDomains in consumer, so just grab
   // and use them.
-  auto root_dom = consumer_tv->getMaybeRFactorDomain();
+  auto alloc_dom = consumer_tv->getMaybeAllocationDomain();
```
Comment (naoyam) [inline csrc/index_compute.cpp:2035]: const auto&

**Review by naoyam — APPROVED** [review]
LGTM

File: csrc/index_compute.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: fixed

File: csrc/index_compute.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: fixed

File: csrc/index_compute.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: fixed

File: csrc/index_compute.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: fixed

**Pull request merged**